handler_test.go

```diff
@@ -0,0 +1,474 @@
+			w.Header().Set("Content-Type", "application/json")
+			w.Write([]byte("true"))
+			return nil
+		}
+	},
+	req: &http.Request{},
+	pathVar: httprouter.Params{{
+		Key:   "a",
+		Value: "123",
+	}},
+	expectBody: true,
+}, {
+	about: "function with error return that returns an error",
+	f: func(c *gc.C) interface{} {
+		type testStruct struct {
+			A int `httprequest:"a,path"`
+		}
+		return func(w http.ResponseWriter, p httprequest.Params, s *testStruct) error {
+			c.Assert(s, jc.DeepEquals, &testStruct{123})
+			return errUnauth
+		}
+	},
+	req: &http.Request{},
+	pathVar: httprouter.Params{{
+		Key:   "a",
+		Value: "123",
+	}},
+	expectBody: errorResponse{
+		Message: errUnauth.Error(),
+	},
+	expectStatus: http.StatusUnauthorized,
+}, {
+	about: "function with value return that returns a value",
+	f: func(c *gc.C) interface{} {
+		type testStruct struct {
+			A int `httprequest:"a,path"`
+		}
+		return func(h http.Header, p httprequest.Params, s *testStruct) (int, error) {
+			c.Assert(s, jc.DeepEquals, &testStruct{123})
+			return 1234, nil
+		}
+	},
+	req: &http.Request{},
+	pathVar: httprouter.Params{{
+		Key:   "a",
+		Value: "123",
+	}},
+	expectBody: 1234,
+}, {
+	about: "function with value return that returns an error",
+	f: func(c *gc.C) interface{} {
+		type testStruct struct {
+			A int `httprequest:"a,path"`
+		}
+		return func(h http.Header, p httprequest.Params, s *testStruct) (int, error) {
+			c.Assert(s, jc.DeepEquals, &testStruct{123})
+			return 0, errUnauth
+		}
+	},
+	req: &http.Request{},
+	pathVar: httprouter.Params{{
+		Key:   "a",
+		Value: "123",
+	}},
+	expectBody: errorResponse{
+		Message: errUnauth.Error(),
+	},
+	expectStatus: http.StatusUnauthorized,
+}, {
+	about: "error when unmarshaling",
+	f: func(c *gc.C) interface{} {
+		type testStruct struct {
+			A int `httprequest:"a,path"`
+		}
+		return func(h http.Header, p httprequest.Params, s *testStruct) (int, error) {
+			c.Errorf("function should not have been called")
+			return 0, nil
+		}
+	},
+	req: &http.Request{},
+	pathVar: httprouter.Params{{
+		Key:   "a",
+		Value: "not a number",
+	}},
+	expectBody: errorResponse{
+		Message: `cannot unmarshal parameters: cannot unmarshal into field: cannot parse "not a number" into int: expected integer`,
+	},
+	expectStatus: http.StatusBadRequest,
+}, {
+	about: "error when unmarshaling single value return",
+	f: func(c *gc.C) interface{} {
+		type testStruct struct {
+			A int `httprequest:"a,path"`
+		}
+		return func(w http.ResponseWriter, p httprequest.Params, s *testStruct) error {
+			c.Errorf("function should not have been called")
+			return nil
+		}
+	},
+	req: &http.Request{},
+	pathVar: httprouter.Params{{
+		Key:   "a",
+		Value: "not a number",
+	}},
+	expectBody: errorResponse{
+		Message: `cannot unmarshal parameters: cannot unmarshal into field: cannot parse "not a number" into int: expected integer`,
+	},
+	expectStatus: http.StatusBadRequest,
+}, {
+	about: "unmarshalable return type",
```

Comment:
Unmarshalable here means not-marshalable, I can see how this could be confusing.